### PR TITLE
Fix route receiving invalid params

### PR DESCRIPTION
**Pull request by mariana-mendes**
#### What is the purpose of this pull request?

as title says.

#### What problem is this solving?
When passing params not specified in [routes template](https://github.com/vtex-apps/store/blob/master/store/routes.json), none of warn is displayed. 

#### How should this be manually tested?
1. [Acess this workspace](https://mariana--storecomponents.myvtex.com)
2. Create a navigate object passing more params then necessary e.g `navigate({page: 'store.product',
  params: { slug: 'traveler-backpack', batata: 'batata', banana: 'banana', a:'a'},})`
3. And check if console throws this warn : 
`pages.ts:116 The following params are invalid:
batata
banana
a`

#### Types of changes

* [X] Bug fix (a non-breaking change which fixes an issue)
* [ ] New feature (a non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/utils/pages.ts [hunk 1]
```diff
@@ -102,12 +109,20 @@ function getRouteFromPageName(
...
 ): NavigationRoute | null {
-  const path = pathFromPageName(id, pages, params)
-  return path ? { id, path, params } : null
+  const path = pathFromPageName(id, pages, params) || ''
+  const template = getValidTemplate(id, pages) || ''
+  const validParams = getParams(template, path) as Record<string, any>
+  const invalidParams = difference(keys(params), keys(validParams))
+
+  if(!isEmpty(invalidParams)){
+    console.warn(`The following params are invalid:\n${invalidParams}\n`)
```
Comment (GustavoMA) [inline react/utils/pages.ts]: use `invalidParams.join('\n')` that will make a better warning message.

File: react/utils/pages.ts [hunk 2]
```diff
@@ -102,12 +109,20 @@ function getRouteFromPageName(
   pages: Pages,
   params: any
 ): NavigationRoute | null {
-  const path = pathFromPageName(id, pages, params)
-  return path ? { id, path, params } : null
+  const path = pathFromPageName(id, pages, params) || ''
+  const template = getValidTemplate(id, pages) || ''
```
Comment (luciannojunior) [inline react/utils/pages.ts]: Extract 113 to 115 in another function

File: react/utils/pages.ts [hunk 3]
```diff
@@ -61,9 +63,13 @@ export function pathFromPageName(page: string, pages: Pages, params: any) {
...
     return null
   }
+  return adjustTemplate(template)
+}
 
-  const properTemplate = adjustTemplate(template)
-  return new RouteParser(properTemplate).reverse(params) || null
+export function pathFromPageName(page: string, pages: Pages, params: any) {
+  const validTemplate = getValidTemplate(page, pages)
+  if(!validTemplate) return null;
```
Comment (luciannojunior) [inline react/utils/pages.ts]:
```suggestion
  if(!validTemplate) return null
```

File: react/utils/pages.ts [hunk 4]
```diff
@@ -61,9 +63,13 @@ export function pathFromPageName(page: string, pages: Pages, params: any) {
...
+  return new RouteParser(validTemplate).reverse(params) || null
```
Comment (luciannojunior) [inline react/utils/pages.ts]: The `|| null` might be unnecessary here. If there's a truthy `validTemplate`, enforced in the line above, I think that there's no way the `reverse` will return falsy

**Comment by luciannojunior** [comment]
Careful with merge conflicts

File: react/utils/pages.ts (hunk 4 above)
Comment (fanny) [reply react/utils/pages.ts]: But this check is just for validTemplate, we still can have errors in params, then null could be necessary isn't?

File: react/utils/pages.ts (hunk 3 above)
Comment (GustavoMA) [reply react/utils/pages.ts]: Actually
```suggestion
  if (!validTemplate) return null
```

File: react/utils/pages.ts [hunk 5]
```diff
@@ -97,13 +103,25 @@ function getPagePath(name: string, pages: Pages) {
   return cname && isHost(cname) ? '/' : pagePath
 }
 
+function getValidParams(id: string, pages: Pages, path: string, params: any) {
+  const template = getValidTemplate(id, pages) || ''
+  const validParams = getParams(template, path) as Record<string, any>
+  const invalidParams = difference(keys(params), keys(validParams))
+
+  if(!isEmpty(invalidParams)){
```
Comment (GustavoMA) [inline react/utils/pages.ts]:
```suggestion
  if (!isEmpty(invalidParams)) {
```

File: react/utils/pages.ts [hunk 6]
```diff
@@ -97,13 +103,25 @@ function getPagePath(name: string, pages: Pages) {
   return cname && isHost(cname) ? '/' : pagePath
 }
 
+function getValidParams(id: string, pages: Pages, path: string, params: any) {
+  const template = getValidTemplate(id, pages) || ''
+  const validParams = getParams(template, path) as Record<string, any>
+  const invalidParams = difference(keys(params), keys(validParams))
+
+  if (!isEmpty(invalidParams)) {
+    console.warn(`The following params are invalid:\n${invalidParams.join('\n')}\n`)
```
Comment (GustavoMA) [inline react/utils/pages.ts]:
```suggestion
    console.warn(`The following params are invalid:\n${invalidParams.join('\n')}`)
```

Comment (GustavoMA) [reply react/utils/pages.ts]: No need for this last breakline :)

File: react/utils/pages.ts (hunk 4 above)
Comment (GustavoMA) [reply react/utils/pages.ts]: @luciannojunior `new RouteParser(validTemplate).reverse(params)` can return `false` as well.
This method was already with `|| null` before this changes.

**Comment by luciannojunior** [comment]
image
 The warning is not working (or shouldn't it work in this case? 🤔 ). Take a look

Nice job so far

**Comment by fanny** [comment]
Hmmm, it's seems that we run of different ways, when i run `$r.navigate({page: 'store.product', params: { slug: 'shirt', nome: 'shirt'}})` the warns are showed

vtex-error

but when i run your commag, i get this error:

vtex-e-rror

**Comment by fanny** [comment]
My `$r` is the `RenderProvider` component

**Pull request merged**